homeassistant/components/octoprint/sensor.py

```diff
@@ -48,29 +47,33 @@ async def async_setup_entry(
 
     assert device_id is not None
 
-    @callback
-    def async_add_tool_sensors(tool):
-        async_add_entities(
-            [
-                OctoPrintTemperatureSensor(
-                    coordinator,
-                    tool.name,
-                    x,
-                    device_id,
-                )
-                for x in ("actual", "target")
-            ]
-        )
-
-    hass.data[DOMAIN][config_entry.entry_id][DATA_UNSUBSCRIBE].append(
-        async_dispatcher_connect(
-            hass, f"{DOMAIN}_{device_id}_new_tool", async_add_tool_sensors
-        )
-    )
+    known_tools = []
 
-    if printer_info := coordinator.data["printer"]:
-        for tool in printer_info.temperatures:
-            async_add_tool_sensors(tool)
+    @callback
+    def async_add_tool_sensors() -> None:
+        for tool in [
+            tool
+            for tool in coordinator.data["printer"].temperatures
+            if tool.name not in known_tools
+        ]:
+            assert device_id is not None
+            known_tools.append(tool.name)
+            async_add_entities(
```

Comment:
Please collect the entities first and then make a single call to `async_add_entities` for this callback.

Reply:
Swapped out to store into a list then make the call